AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1114,26 +1117,61 @@ - (void)__layout
   ASDisplayNodeAssertMainThread();
   ASDN::MutexLocker l(_propertyLock);
   CGRect bounds = self.bounds;
+
+  [self measureNodeWithBoundsIfNecessary:bounds];
+
   if (CGRectEqualToRect(bounds, CGRectZero)) {
     // Performing layout on a zero-bounds view often results in frame calculations
     // with negative sizes after applying margins, which will cause
     // measureWithSizeRange: on subnodes to assert.
     return;
   }
-
+  
   // Handle placeholder layer creation in case the size of the node changed after the initial placeholder layer
   // was created
   if ([self _shouldHavePlaceholderLayer]) {
     [self _setupPlaceholderLayerIfNeeded];
   }
   _placeholderLayer.frame = bounds;
-
+  
   [self layout];
   [self layoutDidFinish];
 }
 
+- (void)measureNodeWithBoundsIfNecessary:(CGRect)bounds
+{
+  // Normally measure will be called before layout occurs. If this doesn't happen, nothing is going to call it at all.
+  // We simply call measureWithSizeRange: using a size range equal to whatever bounds were provided to that element
+  if (self.supernode == nil && !self.supportsRangeManagedInterfaceState && [self _hasDirtyLayout]) {
```

Comment:
We probably want to grab the lock here, at least while checking these three properties.  If we don't, the value of one or more could change as the condition is being evaluated...  There is also locking overhead in each method as it is re-acquired.